src/rail_cmd.cpp

```diff
@@ -1540,6 +1540,80 @@ static Vehicle *UpdateTrainPowerProc(Vehicle *v, void *data)
 	return nullptr;
 }
 
+struct EnsureNoIncompatibleRailtypeTrainOnGroundData {
+	int z;
+	RailType type;
+};
+
+static Vehicle *EnsureNoIncompatibleRailtypeTrainProc(Vehicle *v, void *data)
+{
+	if (v->type != VEH_TRAIN) {
+		if (v->type == VEH_DISASTER || v->type == VEH_AIRCRAFT) return nullptr; // Don't care about flying things
+		return v; // Non-trains are always in the way
```

Comment:
Why would a truck/bus/tram be in the way, and a landed UFO not?
Also, if this returns a vehicle then later on an error about a train being in the way will be given when this could be a truck/bus/tram.

So, might it make sense to just ignore all non-trains? After all, won't a bus/tram be compatible with the new rail type in any case?